sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreRDD.scala

```diff
@@ -29,6 +29,57 @@ import org.apache.spark.sql.internal.SessionState
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.util.SerializableConfiguration
 
+/**
+ * An RDD that allows computations to be executed against [[ReadOnlyStateStore]]s. It
+ * uses the [[StateStoreCoordinator]] to get the locations of loaded state stores
+ * and use that as the preferred locations.
+ */
+class ReadOnlyStateStoreRDD[T: ClassTag, U: ClassTag](
+    dataRDD: RDD[T],
+    storeReadFunction: (ReadOnlyStateStore, Iterator[T]) => Iterator[U],
+    checkpointLocation: String,
+    queryRunId: UUID,
+    operatorId: Long,
+    storeVersion: Long,
+    keySchema: StructType,
+    valueSchema: StructType,
+    indexOrdinal: Option[Int],
+    sessionState: SessionState,
+    @transient private val storeCoordinator: Option[StateStoreCoordinatorRef],
+    extraOptions: Map[String, String] = Map.empty) extends RDD[U](dataRDD) {
+
+  private val storeConf = new StateStoreConf(sessionState.conf, extraOptions)
+
+  // A Hadoop Configuration can be about 10 KB, which is pretty big, so broadcast it
+  private val hadoopConfBroadcast = dataRDD.context.broadcast(
+    new SerializableConfiguration(sessionState.newHadoopConf()))
+
+  override protected def getPartitions: Array[Partition] = dataRDD.partitions
+
+  /**
+   * Set the preferred location of each partition using the executor that has the related
+   * [[StateStoreProvider]] already loaded.
+   */
+  override def getPreferredLocations(partition: Partition): Seq[String] = {
+    val stateStoreProviderId = StateStoreProviderId(
+      StateStoreId(checkpointLocation, operatorId, partition.index),
+      queryRunId)
+    storeCoordinator.flatMap(_.getLocation(stateStoreProviderId)).toSeq
+  }
+
+  override def compute(partition: Partition, ctxt: TaskContext): Iterator[U] = {
+    val storeProviderId = StateStoreProviderId(
+      StateStoreId(checkpointLocation, operatorId, partition.index),
+      queryRunId)
+
+    val store = StateStore.getReadOnly(
+      storeProviderId, keySchema, valueSchema, indexOrdinal, storeVersion,
```

Comment:
Is this only difference from `StateStoreRDD`? Looks they are almost duplicate.

Reply:
The returning type is different between twos and the type of state function is also different. That's the main intention of the new changes - stateReadFunction never be able to do modification of the state as it will get ReadOnlyStateStore, whereas stateUpdateFunction can.

I agree lots of code are duplicated - that's why I extracted abstract class, and as you saw, MiMa complained. I don't know why MiMa complained as I don't see any weird in previous code but I might be missing so I had to make MiMa happy for now.

Reply:
Just dealt it with 5c70db0 .

 `close` method throws UnsupportedException by default, and callers are catching it. The change considers backward compatibility like below:

1. 
If the implementation of StateStore isn't aware of `close` method, they should have implemented the resource cleanup in `commit` and `abort`, and Spark will call it to clean up resource. Spark will also call `close` which will be no-op. 

For read-only state store instance, they would use WrappedReadStateStore which calls both `abort` and `close` in `close` method, and `abort` should handle the resource cleanup.

2. 
If the implementation of StateStore is aware of `close` method and implements the method correctly, they should have migrated the resource cleanup to `close` method, and Spark will call it to clean up resource. 

For read-only state store instance, they may override `StateStoreProvider.getReadOnlyStore` (as HDFSBackedStateStoreProvider does) to provide its own ReadOnlyStore implementation. Or, they may still use WrappedReadStateStore which calls both `abort` and `close` in `close` method. `abort` would have probably no-op after migrating resource cleanup to `close`, and `close` should handle the resource cleanup.

If this sounds too complicated, let's revert 5c70db0 and live with `abort` method.

Reply:
Sounds too complicated, and it's also a bit weird and read-only store calls both abort and close. Let's keep with original abort and commit for now.

Reply:
OK just reverted 5c70db0.